Significantly decrease MemoryTracking drift

azat opened the pull request:
Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Use total_memory_tracker when there is no other MemoryTracker object (this should significantly reduce the MemoryTracking drift)
- Fix parent memory tracker during query detaching
- Fix accounting for new/delete from different threads for VariableContext::Thread (supersedes #16122)

Detailed description / Documentation draft:
**After this PR the memory drift is ~0 after running 100 queries via http/tcp/tcp from one session. ([test](https://github.com/ClickHouse/ClickHouse/pull/16121/files#diff-70d939d0ce5a53be0dd9313cfeb0c805aaaa2fe8920e1f7e67cdc56bfd42c8ea))**

TL;DR;

<details>

To track memory, clickhouse creates memory tracker object for each
thread **explicitly**, but until it is not created the memory
allocations are not under account.
There should not be lot of allocations w/o memory tracker, since most of
the time it is created early enough, but even this maybe enough to
trigger some problems.

Plus sometimes it is not possible to create it, for example some 3d
party library does not allow to do this explicitly:
- for example before #15740 allocations from librdkafka threads,
- or even worse, poco threads, they don't have any routines to do this.
This won't be a problem for `MemoryTracking` metric if the deallocation
will be done from the same thread w/o memory tracker (or vise versa),
but this is not always true.

NOTE, that this will slow down per-thread allocations w/o memory
tracker, since before this patch there were no memory tracking for them
while now they will be accounted in total_memory_tracker, and for
total_memory_tracker max_untracked_memory is always reached.
But this should not be significant.

</details>

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

<details>

HEAD:
- 7081b1047d50d403ae6ab4935f253107196464d9
- 990ad21b05254feae81341289e188d2d202b1ed1
- 43dd401ae5bd99ac954ad3d574c8f368c779b4f0 - worked, but problem in test
- 327d822b7f1b46b00d9f6167f389580a0ebbac0b - dump attempt to make the test pass

</details>

Fixes: #15932

azat committed:
Use total_memory_tracker when there is no other MemoryTracker object.

This should significantly reduce the MemoryTracking drift, test shows
that there is 0 drift after query storm (100 queries, via http/tcp/tcp
in one session).

TL;DR;

To track memory, clickhouse creates memory tracker object for each
thread **explicitly**, but until it is not created the memory
allocations are not under account.
There should not be lot of allocations w/o memory tracker, since most of
the time it is created early enough, but even this maybe enough to
trigger some problems.

Plus sometimes it is not possible to create it, for example some 3d
party library does not allow to do this explicitly:
- for example before #15740 allocations from librdkafka threads,
- or even worse, poco threads, they don't have any routines to do this.
This won't be a problem for `MemoryTracking` metric if the deallocation
will be done from the same thread w/o memory tracker (or vise versa),
but this is not always true.

NOTE, that this will slow down per-thread allocations w/o memory
tracker, since before this patch there were no memory tracking for them
while now they will be accounted in total_memory_tracker, and for
total_memory_tracker max_untracked_memory is always reached.
But this should not be significant.

azat committed:
Fix accounting for new/delete from different threads for VariableContext::Thread

MemoryTracker assumes that for VariableContext::Thread new/delete may be
called from different threads, hence the amount of memory can go
negative.

However the MemoryTracker is nested, so even if the negative amount is
allowed for VariableContext::Thread it does not allowed for anything
upper, and hence the MemoryTracking will not be decremented properly.

Fix this, by passing initial size to the parent free.

This should fix memory drift for HTTP queries.

azat committed:
Add a test for memory drift in user memory tracker (max_memory_usage_for_user)

azat committed:
Add a test for MemoryTracking drift

v2: disable query profiling and logging in 01540_MemoryTracking
    (This should make MemoryTracker drift zero).

alexey-milovidov commented on src/Common/MemoryTracker.cpp:
```diff
@@ -6,13 +6,40 @@
+        return thread_memory_tracker;
+
+    /// Do not use total_memory_tracker automatically in the main thread:
+    /// - there is not too much allocations in the main thread,
```
\- there is not too much allocations
\+ there are not too many allocations

azat replied:
Actually it will not be used there anyway, since it has it's own, so this comment will go away anyway

alexey-milovidov commented:
I like the idea, I have also tried to make something like that but then found it difficult.

azat commented:
>I like the idea, I have also tried to make something like that but then found it difficult.

Ok, great (also it worth add a message when it is synced with RSS to see how it will differs later)

It still in draft due to gettid usage (that fact that performance does not shows the regression does not mean that there is no regression in this case, since most of the time total memory tracker is not used)

alexey-milovidov commented:
You can also compare addresses of dummy thread_local variable (with the address of the same var from main).

alexey-milovidov commented:
Also you can use atomic bool in constructor of thread-local variable and check it for non-zero.

azat commented:
Manual testing shows significant improvements, for http/tcp queries storm, I guess mosty due to memory accounting from the poco thread pools (require some test, will add later)

azat committed:
Fix parent memory tracker during query detaching

azat commented:
Added a test that checks that after 100 queries the MemoryTracking does not changed (even to 1 byte) more then 6% (6 times), and it passed now (while before the metric is changed, and actually grows in some cases to ~4MB, after each query)!

akuzm commented on tests/queries/0_stateless/01540_MemoryTracking.sh:
```diff
@@ -0,0 +1,96 @@
+
+# wait until the server will be initialized
+# (since it may do some costly allocations)
+sleep 1m
```
Looks like this is the first line of the script that is ran, so what is it waiting for?

azat replied:
Yes, this is just an attempt in background, while I was digging into the issue (that was `log_queries`), this patch will be removed shortly.

azat committed:
Make 01540_MemoryTracking integration

01540_MemoryTracking is failing on CI for the following reasons:
- log_queries (fixed, by adding log_queries=0)
- profilers (fixed)
- but what can't be fixed is metric_log and so on, so we need separate
  instance with separate configuration (sigh).

azat commented:
So the problem was only in the test

Due to metric_log it has some difference, of course I could rewrite the test to allow some small delta, ~20K is enough, but I want ensure/show that after some actions the MemoryTracking does not changes at all.
So the stateless test was replaced with integration test (basically it is needed only to run instance of the server w/o `system.*_log`)

azat committed:
Disable syncing MemoryTracking with RSS for test_MemoryTracking

azat committed:
Tune TTL of the background query in 01541_max_memory_usage_for_user

azat commented:
So now looks OK (although there is minimal conflict in the fasttest skip list, it is pretty trivial so I will not solve it by myself to avoid trigerring CI)

alexey-milovidov committed:
Merge branch 'master' into total_memory_tracker-by-default

alexey-milovidov commented on src/Common/MemoryTracker.cpp:
```diff
@@ -192,14 +210,15 @@ void MemoryTracker::free(Int64 size)
         DB::TraceCollector::collect(DB::TraceType::MemorySample, StackTrace(), -size);
     }
 
+    Int64 accounted_size = size;
```
Don't understand this change.

alexey-milovidov replied:
Ok.

alexey-milovidov commented on src/Interpreters/ThreadStatusExt.cpp:
```diff
@@ -301,7 +301,7 @@ void ThreadStatus::detachQuery(bool exit_if_already_detached, bool thread_exits)
 
     /// Must reset pointer to thread_group's memory_tracker, because it will be destroyed two lines below.
-    memory_tracker.setParent(nullptr);
+    memory_tracker.setParent(thread_group->memory_tracker.getParent());
```
Commend becomes obsolete.

alexey-milovidov committed:
Update ThreadStatusExt.cpp
